common/bucket_utils.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions for interacting with the buckets."""
+
+from common import logs
+from common import experiment_utils
+
+logger = logs.Logger('bucket_utils')
+
+
+def __using_gsutil():
+    """ Test whether using gsutil format paths. """
+    try:
+        experiment_path_format = experiment_utils.get_cloud_experiment_path()
```

Comment:
I'm not sure why this will be empty in a local experiment care to explain?

Reply:
`get_cloud_experiment_path` will check the env. And for some test cases, there is no env setting before calling bucket_utils. Then it will raise a KeyError.

Reply:
What are those cases?

Reply:
Here are the 7 error messages if there is no exception catching logic:
```
============================= ERRORS ==============================
_________ ERROR collecting experiment/test_dispatcher.py __________
experiment/test_dispatcher.py:25: in <module>
    from experiment import dispatcher
experiment/dispatcher.py:33: in <module>
    from experiment.build import builder
experiment/build/builder.py:32: in <module>
    from experiment.build import build_utils
experiment/build/build_utils.py:19: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
__________ ERROR collecting experiment/test_measurer.py ___________
experiment/test_measurer.py:27: in <module>
    from experiment.build import build_utils
experiment/build/build_utils.py:19: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
__________ ERROR collecting experiment/test_reporter.py ___________
experiment/test_reporter.py:19: in <module>
    from experiment import reporter
experiment/reporter.py:21: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
_______ ERROR collecting experiment/test_run_experiment.py ________
experiment/test_run_experiment.py:23: in <module>
    from experiment import run_experiment
experiment/run_experiment.py:33: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
___________ ERROR collecting experiment/test_runner.py ____________
experiment/test_runner.py:23: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
________ ERROR collecting experiment/build/test_builder.py ________
experiment/build/test_builder.py:23: in <module>
    from experiment.build import builder
experiment/build/builder.py:32: in <module>
    from experiment.build import build_utils
experiment/build/build_utils.py:19: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
____ ERROR collecting service/test_automatic_run_experiment.py ____
service/test_automatic_run_experiment.py:19: in <module>
    from service import automatic_run_experiment
service/automatic_run_experiment.py:29: in <module>
    from experiment import run_experiment
experiment/run_experiment.py:33: in <module>
    from common import bucket_utils
common/bucket_utils.py:28: in <module>
    if _using_gsutil():
common/bucket_utils.py:24: in _using_gsutil
    experiment_path_format = experiment_utils.get_cloud_experiment_path()
common/experiment_utils.py:47: in get_cloud_experiment_path
    cloud_experiment_bucket = os.environ['CLOUD_EXPERIMENT_BUCKET']
/usr/lib/python3.7/os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'CLOUD_EXPERIMENT_BUCKET'
!!!!!!!!!!!!! Interrupted: 7 errors during collection !!!!!!!!!!!!!
======================== 7 errors in 2.75s ========================
Failed.
make: *** [Makefile:28: presubmit] Error 1

```

Reply:
So then those tests need to be fixed. It's bad practice to add test only code to production code. I think all of these can be fixed by adding the `experiment` fixture as a param, or by loading the module lazily.

Reply:
From another view, unit test are usually not acquiring more information than that module itself. `experiment_utils.get_cloud_experiment_path()` is already a higher level function, not independent to `filestore_utils` itself. Thus, we may need provide some default behavior for this component.

Reply:
By the way, there will be another env called `experiment_path` to use in the next PR.